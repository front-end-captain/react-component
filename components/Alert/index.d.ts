import * as React from "react";

interface alertProps {
  type?: "info" | "warn" | "success" | "error";
  closable?: boolean;
  title?: React.ReactNode;
  content?: React.ReactNode;
  onClose?: React.MouseEventHandler<HTMLAnchorElement>;
  afterClose?: () => void;
  showIcon?: boolean;
  icon?: React.ReactNode;
  style?: React.CSSProperties;
  className?: string;
  prefixCls?: string;
}

interface alertState {
  closing: boolean;
  closed: boolean;
}

export default class Alert extends React.Component<alertProps, alertState> {}
