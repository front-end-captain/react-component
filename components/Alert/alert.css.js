import styled from "styled-components";

import { skyAlertHide } from "./../_style/animate.js";

const AlertWrapper = styled.div`
  display: flex;
  flex-direction: row;
  font-size: 14px;
  font-feature-settings: "tnum";
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  position: relative;
  font-variant: tabular-nums;
  margin: 0;
  list-style: none;
  padding: 8px 15px 8px 16px;
  border-radius: 4px;

  &.${(props) => props.prefixCls}-closing {
    animation: ${skyAlertHide} 0.5s ease-out;
    animation-fill-mode: both;
    transform-origin: left top;
  }

  &.${(props) => props.prefixCls}-show-icon {
    padding: 8px 30px 8px 14px;
  }

  &.${(props) => props.prefixCls}-no-content {
    .${(props) => props.prefixCls}-title {
      margin-bottom: 0 !important;
    }

    .${(props) => props.prefixCls}-icon {
      align-items: center;
    }
  }

  &.${(props) => props.prefixCls}-no-title {
    .${(props) => props.prefixCls}-icon {
      font-size: 14px;
      align-items: center;
      padding-top: 0;
    }

    .${(props) => props.prefixCls}-container {
      flex: 14 1;
    }
  }

  &.${(props) => props.prefixCls}-success {
    background-color: rgba(82, 196, 26, 0.5);
    border: 1px solid rgb(82, 196, 26);

    .${(props) => props.prefixCls}-icon {
      color: rgb(82, 196, 26);
    }
  }

  &.${(props) => props.prefixCls}-info {
    background-color: rgba(19, 194, 194, 0.5);
    border: 1px solid rgb(19, 194, 194);

    .${(props) => props.prefixCls}-icon {
      color: rgb(19, 194, 194);
    }
  }

  &.${(props) => props.prefixCls}-warn {
    background-color: rgba(250, 140, 22, 0.5);
    border: 1px solid rgb(250, 140, 22);

    .${(props) => props.prefixCls}-icon {
      color: rgb(250, 140, 22);
    }
  }

  &.${(props) => props.prefixCls}-error {
    background-color: rgba(245, 34, 45, 0.5);
    border: 1px solid rgb(245, 34, 45);

    .${(props) => props.prefixCls}-icon {
      color: rgb(245, 34, 45);
    }
  }

  .${(props) => props.prefixCls}-icon {
    flex: 1 1;
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    font-size: 24px;
    padding-top: 2px;
  }

  .${(props) => props.prefixCls}-container {
    display: flex;
    flex-direction: column;
    flex: 10 1;

    .${(props) => props.prefixCls}-title {
      font-size: 16px;
      color: rgba(0, 0, 0, 0.85);
      display: block;
      margin-bottom: 4px;
    }

    .${(props) => props.prefixCls}-content {
      font-size: 14px;
      line-height: 22px;
    }
  }

  .${(props) => props.prefixCls}-close-icon {
    position: absolute;
    font-size: 14px;
    right: 12px;
    top: 10px;
    line-height: 22px;
    cursor: pointer;
    overflow: hidden;
  }
`;

export default AlertWrapper;
