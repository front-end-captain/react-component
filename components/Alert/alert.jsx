import React, { Component, isValidElement, cloneElement } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { InfoIcon, SuccessIcon, ErrorIcon, WarningIcon, CloseIcon } from "./../Icon/index.jsx";

import AlertWrapper from "./alert.css.js";

const typeMap = {
  info: "info",
  warn: "warn",
  success: "success",
  error: "error",
};

class Alert extends Component {
  static defaultProps = {
    prefixCls: "sky-alert",
    type: "info",
    closable: false,
    showIcon: false,
    onClose: () => {},
    afterClose: () => {},
  };

  static propTypes = {
    prefixCls: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.objectOf(PropTypes.string),
    type: PropTypes.oneOf(Object.values(typeMap)),
    closable: PropTypes.bool,
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    content: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    onClose: PropTypes.func,
    afterClose: PropTypes.func,
    showIcon: PropTypes.bool,
    icon: PropTypes.element,
  };

  state = {
    closing: false,
    closed: false,
  };

  animationTime = 500;

  getIcon = () => {
    const { icon, prefixCls, type } = this.props;

    const iconClassName = `${prefixCls}-icon`;

    if (isValidElement(icon)) {
      return cloneElement(icon, { className: iconClassName });
    }

    if (type === typeMap.success) {
      return <SuccessIcon className={iconClassName} />;
    }

    if (type === typeMap.warn) {
      return <WarningIcon className={iconClassName} />;
    }

    if (type === typeMap.error) {
      return <ErrorIcon className={iconClassName} />;
    }

    return <InfoIcon className={iconClassName} />;
  };

  handleClose = (event) => {
    const { onClose, afterClose } = this.props;

    this.setState({ closing: true }, () => {
      onClose(event);

      setTimeout(() => {
        this.setState({ closed: true, closing: false }, () => {
          afterClose();
        });
      }, this.animationTime);
    });
  };

  render() {
    const { closed, closing } = this.state;
    const { title, content, prefixCls, className, style, type, showIcon, closable } = this.props;

    const wrapperClassName = classNames(`${prefixCls}`, className, `${prefixCls}-${type}`, {
      [`${prefixCls}-show-icon`]: showIcon,
      [`${prefixCls}-no-content`]: !content,
      [`${prefixCls}-no-title`]: !title,
      [`${prefixCls}-closing`]: closing,
    });

    if (closed) {
      return null;
    }

    return (
      <AlertWrapper className={wrapperClassName} style={style} prefixCls={prefixCls}>
        {showIcon && (
          <div className={`${prefixCls}-icon`}>
            <span>{this.getIcon()}</span>
          </div>
        )}
        <div className={`${prefixCls}-container`}>
          {title && <span className={`${prefixCls}-title`}>{title}</span>}
          {content && <span className={`${prefixCls}-content`}>{content}</span>}
        </div>
        {closable && (
          <div className={`${prefixCls}-close-icon`}>
            <span onClick={this.handleClose}>
              <CloseIcon />
            </span>
          </div>
        )}
      </AlertWrapper>
    );
  }
}

export default Alert;
