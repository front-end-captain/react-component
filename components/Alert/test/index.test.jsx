import React from "react";
import toJson from "enzyme-to-json";
import { render, shallow } from "enzyme";

import Alert from "./../index.js";

describe("Alert", () => {
  it("should be render some Alert components", () => {
    const wrapper = render(
      <div>
        <Alert content="some content" />
        <Alert title="Title" content="some content" type="warn" />
        <Alert closable content="some content" type="error" />
        <Alert showIcon content="some content" type="success" />
        <Alert showIcon closable title="This is title" content="some content" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should invoke external callback correctly", () => {
    const onClose = jest.fn();
    const afterClose = jest.fn();

    const wrapper = shallow(
      <Alert onClose={onClose} afterClose={afterClose} closable content="some content" />,
    );

    wrapper.find(".sky-alert-close-icon > span").simulate("click");

    expect(onClose).toHaveBeenCalled();
    setTimeout(() => {
      expect(afterClose).toHaveBeenCalled();
    }, 500);
  });

  it("should render custom icon when it's set icon", () => {
    const customIcon = <div className="custom-indicator" />;
    const wrapper = render(<Alert icon={customIcon} />);
    expect(wrapper).toMatchSnapshot();
  });
});
