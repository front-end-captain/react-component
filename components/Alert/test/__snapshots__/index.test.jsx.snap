// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should be render some Alert components 1`] = `
<div>
  <div
    class="sky-alert sky-alert-info sky-alert-no-title alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
  >
    <div
      class="sky-alert-container"
    >
      <span
        class="sky-alert-content"
      >
        some content
      </span>
    </div>
  </div>
  <div
    class="sky-alert sky-alert-warn alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
  >
    <div
      class="sky-alert-container"
    >
      <span
        class="sky-alert-title"
      >
        Title
      </span>
      <span
        class="sky-alert-content"
      >
        some content
      </span>
    </div>
  </div>
  <div
    class="sky-alert sky-alert-error sky-alert-no-title alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
  >
    <div
      class="sky-alert-container"
    >
      <span
        class="sky-alert-content"
      >
        some content
      </span>
    </div>
    <div
      class="sky-alert-close-icon"
    >
      <span>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="sky-alert sky-alert-success sky-alert-show-icon sky-alert-no-title alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
  >
    <div
      class="sky-alert-icon"
    >
      <span>
        <svg
          class="sky-alert-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 48c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m140.204 130.267l-22.536-22.718c-4.667-4.705-12.265-4.736-16.97-.068L215.346 303.697l-59.792-60.277c-4.667-4.705-12.265-4.736-16.97-.069l-22.719 22.536c-4.705 4.667-4.736 12.265-.068 16.971l90.781 91.516c4.667 4.705 12.265 4.736 16.97.068l172.589-171.204c4.704-4.668 4.734-12.266.067-16.971z"
          />
        </svg>
      </span>
    </div>
    <div
      class="sky-alert-container"
    >
      <span
        class="sky-alert-content"
      >
        some content
      </span>
    </div>
  </div>
  <div
    class="sky-alert sky-alert-info sky-alert-show-icon alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
  >
    <div
      class="sky-alert-icon"
    >
      <span>
        <svg
          class="sky-alert-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M235.4 172.2c0-11.4 9.3-19.9 20.5-19.9 11.4 0 20.7 8.5 20.7 19.9s-9.3 20-20.7 20c-11.2 0-20.5-8.6-20.5-20zm1.4 35.7H275V352h-38.2V207.9z"
          />
          <path
            d="M256 76c48.1 0 93.3 18.7 127.3 52.7S436 207.9 436 256s-18.7 93.3-52.7 127.3S304.1 436 256 436c-48.1 0-93.3-18.7-127.3-52.7S76 304.1 76 256s18.7-93.3 52.7-127.3S207.9 76 256 76m0-28C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
          />
        </svg>
      </span>
    </div>
    <div
      class="sky-alert-container"
    >
      <span
        class="sky-alert-title"
      >
        This is title
      </span>
      <span
        class="sky-alert-content"
      >
        some content
      </span>
    </div>
    <div
      class="sky-alert-close-icon"
    >
      <span>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Alert should render custom icon when it's set icon 1`] = `
<div
  class="sky-alert sky-alert-info sky-alert-no-content sky-alert-no-title alertcss__AlertWrapper-sc-1m63gp0-0 iujTdz"
>
  <div
    class="sky-alert-container"
  />
</div>
`;
